keps/sig-node/3288-separate-stdout-from-stderr/README.md

```diff
@@ -0,0 +1,807 @@
+	// Otherwise, the combined stdout and stderr from the container is returned.
+	// Available values are: "stdout", "stderr", "" (empty string).
+    Stream string
+}
+```
+
+When users want to query certain stream from container, they need to add a new query named `stream`
+to the URL, i.e. `/api/v1/namespaces/default/pods/foo/log?stream=stderr&container=nginx`.
+Then the kube-apiserver is able to know the desired `stream` and passes it to the kubelet.
+
+To tell kubelet which stream to return, we need to update the `LogLocation` function to make it be aware of
+the new parameter:
+
+```go
+// LogLocation returns the log URL for a pod container. If opts.Container is blank
+// and only one container is present in the pod, that container is used.
+func LogLocation(
+	ctx context.Context, getter ResourceGetter,
+	connInfo client.ConnectionInfoGetter,
+	name string,
+	opts *api.PodLogOptions,
+) (*url.URL, http.RoundTripper, error) {
+	...
+    params := url.Values{}
+    // validate stream
+    switch opts.Stream {
+        case "stdout", "stderr", "":
+        default:
+            return nil, nil, errors.NewBadRequest(fmt.Sprintf("invalid container log stream %s", opts.Stream))
+    }		
+    params.Add("stream", opts.Stream)
+	...
+}
+```
+
+### Changes of kubelet
+
+Add a new field `Stream` to `k8s.io/api/core/v1.PodLogOptions`:
+
+```go
+// LogStreamType represents the desired log stream type.
+// +enum
+type LogStreamType string
+
+const (
+    // LogStreamTypeStdout is the stream type for stdout.
+    LogStreamTypeStdout LogStreamType = "stdout"
+    // LogStreamTypeStderr is the stream type for stderr.
+	LogStreamTypeStderr LogStreamType = "stderr"
+    // LogStreamTypeAll represents the combined stdout and stderr.
+	LogStreamTypeAll    LogStreamType = ""
+)
+// PodLogOptions is the query options for a Pod's logs REST call.
+type PodLogOptions struct {
+...
+    // If set, return the given log stream of the container.
+    // Otherwise, the combined stdout and stderr from the container is returned.
+    // +optional
+    Stream LogStreamType
+}
+```
+
+In the `getContainerLogs` method of `k8s.io/kubernetes/pkg/kubelet/server.Server`, we examine the `Stream` field
+of `PodLogOptions` to decide which stream to return:
+
+```go
+// getContainerLogs handles containerLogs request against the Kubelet
+func (s *Server) getContainerLogs(request *restful.Request, response *restful.Response) {
+...
+    fw := flushwriter.Wrap(response.ResponseWriter)
+    var stdout, stderr io.Writer
+	switch logOptions.Stream {
+        case corev1.LogStreamTypeStdout:
+			stdout, stderr = fw, io.Discard
+        case corev1.LogStreamTypeStderr:
+			stdout, stderr = io.Discard, fw
+        case corev1.LogStreamTypeAll:
+			stdout, stderr = fw, fw
+    }
+    response.Header().Set("Transfer-Encoding", "chunked")
+    if err := s.host.GetKubeletContainerLogs(ctx, kubecontainer.GetPodFullName(pod), containerName, logOptions, stdout, stderr); err != nil {
```

Comment:
So output is already captured separately, i.e. this `s.host.GetKubeletContainerLog` is really able to return just stderr or stdout?

How does this currently handle interleaving when a single stream is requested?

I was under the impression that stdout and stderr was redirected to the same pipe to achieve that, in which case splitting it later won't be possible. But perhaps there are runtimes which handle this better or I don't remember correctly?

Reply:
I checked again and at least crio uses separate pipes, so it could be possible to pass that data on separately.

Reply:
> this s.host.GetKubeletContainerLog is really able to return just stderr or stdout?

@pohly Yeah `s.host.GetKubeletContainerLogs` eventually invokes [`ReadLogs`](https://github.com/kubernetes/kubernetes/blob/d45cff711648fcbf597780864da8239db7f3f341/pkg/kubelet/kuberuntime/logs/logs.go#L281), which will return respective stream: https://github.com/kubernetes/kubernetes/blob/d45cff711648fcbf597780864da8239db7f3f341/pkg/kubelet/kuberuntime/logs/logs.go#L252-L258

Reply:
BTW, we can assume that the container runtime is CRI compatible (including its logs).

Reply:
Looks good, thanks for explaining.

I would "resolve" this comment thread, but GitHub doesn't let me.

Reply:
@pohly Thanks for your review 🚀 I could resolve this thread 😉